test/unit/core/test-resource-container-helper.js

```diff
@@ -0,0 +1,196 @@
+    ampParent = createAmpElement({id: 'parent', parent: ampGrandparent});
+    ampParentSibling = createAmpElement({id: 'parent-sibling', parent: ampGrandparent});
+    ampChild1 = createAmpElement({id: 'child1', parent: ampParent});
+    ampChild2 = createAmpElement({id: 'child2', parent: ampParent});
+  });
+
+  function createAmpElement({id, parent, createPlaceholder = true}) {
+    const element = doc.createElement('div');
+    element.id = id;
+    element.classList.add('i-amphtml-element');
+
+    if (createPlaceholder) {
+      const placeholder = doc.createElement('div');
+      placeholder.setAttribute('placeholder', '');
+      element.appendChild(placeholder);
+      const placeholderAmpChild = createAmpElement({id: `placeholder-${id}`, parent: placeholder, createPlaceholder: false});
+      element.getPlaceholder = () => placeholder;
+      element.placeholderAmpChild = placeholderAmpChild;
+      element.placeholderAmpGrandchild = createAmpElement({id: `placeholder-child-${id}`, parent: placeholderAmpChild, createPlaceholder: false});
+    } else {
+      element.getPlaceholder = () => null;
+    }
+
+    element.ensureLoaded = env.sandbox.stub();
+    element.pause = env.sandbox.stub();
+    element.unmount = env.sandbox.stub();
+
+    parent.appendChild(element);
+    return element;
+  }
+
+  describe('forAllWithin', () => {
+    let callback;
+
+    beforeEach(() => {
+      callback = env.sandbox.stub();
+    });
+
+    it('should include self in shallow mode and include placeholder', () => {
+      forAllWithin(ampGrandparent, INCLUDE_SELF, !DEEP, callback);
+      expect(callback).to.be.calledTwice
+        .calledWith(ampGrandparent)
+        .calledWith(ampGrandparent.placeholderAmpChild);
+      // Shouldn't be called for a placeholder grandchild.
+      expect(callback).to.not.be.calledWith(ampGrandparent.placeholderAmpGrandchild);
+    });
+
+    it('should only include the nearest first element in includeSelf/shallow mode', () => {
+      forAllWithin(tree, INCLUDE_SELF, !DEEP, callback);
+      expect(callback).to.be.calledTwice
+        .calledWith(ampGrandparent)
+        .calledWith(ampGrandparentSibling);
+    });
+
+    it('should only include shallow elements in shallow mode not including self', () => {
+      forAllWithin(ampGrandparent, !INCLUDE_SELF, !DEEP, callback);
+      expect(callback).to.be.calledThrice
+        .calledWith(ampParent)
+        .calledWith(ampParentSibling)
+        .calledWith(ampGrandparent.placeholderAmpChild);
+    });
+
+    it('should include self in deep mode', () => {
+      forAllWithin(ampParent, INCLUDE_SELF, DEEP, callback);
+      expect(callback).to.not.be.calledWith(ampGrandparent);
+      expect(callback).to.be.calledWith(ampParent);
+      expect(callback).to.be.calledWith(ampChild1);
+      expect(callback).to.be.calledWith(ampChild2);
+
+      // And placeholders.
+      expect(callback).to.be.calledWith(ampParent.placeholderAmpChild);
```

Comment:
Should we also expect to be called with `ampChild1.placeholderAmpChild` and `ampChild2.placeholderAmpChild`?

Reply:
Yes. But it seemed like superfluous, since at this level they are simply descendants of the parent like any other AMP elements.